server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java

```diff
@@ -830,4 +845,73 @@ public Void withHandle(Handle handle)
       log.warn(e, "Exception while deleting records from table");
     }
   }
+
+  /**
+   * create index on the table if not already exist with retry.
+   *
+   * @param handle Connection handle
+   * @param table  name of the table to fetch the index map
+   * @return Map of the uppercase index name to List of the columns in index
+   */
+  public Map<String, List<String>> getIndexOnTable(Handle handle, String table)
+  {
+    Map<String, List<String>> res = new HashMap<>();
+    try {
+      DatabaseMetaData databaseMetaData = handle.getConnection().getMetaData();
+      ResultSet resultSet = databaseMetaData.getIndexInfo(null, null, table.toUpperCase(Locale.ENGLISH), false, false);
+
+      while (resultSet.next()) {
+        String indexName = resultSet.getString("INDEX_NAME");
+        String columnName = resultSet.getString("COLUMN_NAME");
+        if (org.apache.commons.lang.StringUtils.isNotBlank(indexName) && org.apache.commons.lang.StringUtils.isNotBlank(
+            columnName)) {
+          columnName = StringUtils.toUpperCase(columnName);
+          List<String> cols = res.getOrDefault(indexName, new ArrayList<>());
+          cols.add(columnName);
+          res.put(indexName, cols);
+        }
+      }
+    }
+    catch (Exception e) {
+      log.error(e, "Exception while listing the index on table %s ", table);
+    }
+    return res;
+  }
+
+  /**
+   * create index on the table if not already exist with retry.
+   *
+   * @param tableName   Name of the table to create index on
+   * @param checkString case-insensitive string which substring of index name, it helps to check the existing index on table
+   * @param indexSQL    SQL statement to create index
+   */
+  public void createIndex(final String tableName, final String checkString, final String indexSQL)
+  {
+    try {
+      retryWithHandle(
+          new HandleCallback<Void>()
+          {
+            @Override
+            public Void withHandle(Handle handle)
+            {
+              if (tableExists(handle, tableName)) {
+                Map<String, List<String>> indexMap = getIndexOnTable(handle, tableName);
```

Comment:
Your function signatures are going to cause us to read all of the indexes for all of the tables on every single check.  Not a big deal for 2, but better to be more conscious of the number of DB calls that you are making.

If you adjust the signature of `createIndex` to be a bit more developer friendly (as suggested above), then I think that you can build the exact index name that you expect and check for only that.  Alternatively, you should read the set of indexes once and then reuse that map for each of these calls.